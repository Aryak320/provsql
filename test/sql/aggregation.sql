\set ECHO none
SET search_path TO public,provsql;

--aggregation formula

CREATE FUNCTION public.formula_semimod(formula1 text, formula2 text) RETURNS text
    LANGUAGE sql IMMUTABLE STRICT
    AS $$
  SELECT concat('(',formula1,' * ',formula2,')')
$$;

CREATE FUNCTION public.formula_agg_state(state public.formula_state, value text) RETURNS public.formula_state
    LANGUAGE plpgsql IMMUTABLE
    AS $$
BEGIN
  IF state IS NULL OR state.nbargs=0 THEN
    RETURN (value,1);
  ELSE
    RETURN (concat(state.formula,' , ',value),state.nbargs+1);
  END IF;
END
$$;

CREATE AGGREGATE public.formula_agg(text) (
    SFUNC = public.formula_agg_state,
    STYPE = public.formula_state,
    INITCOND = '(1,0)'
);

CREATE FUNCTION public.formula_agg_final(state public.formula_state, fname varchar) RETURNS text
  LANGUAGE sql IMMUTABLE STRICT
  AS
  $$
    SELECT concat(fname,'{ ',state.formula,' }');
  $$;

CREATE FUNCTION public.aggregation_formula(token provsql.provenance_token, token2value regclass) RETURNS text
    LANGUAGE plpgsql
    AS $$
BEGIN
  RETURN aggregation_evaluate(
    token,
    token2value,
    'formula_agg_final',
    'formula_agg',
    'formula_semimod',
    'ùüô'::text,
    'formula_plus',
    'formula_times',
    'formula_monus',
    'formula_delta');
END
$$;

CREATE TABLE agg_result AS
    SELECT position, count(*), formula(provenance(),'personnel_name') FROM personnel
    GROUP BY position;

SELECT remove_provenance('agg_result');

SELECT * FROM agg_result ORDER BY position;

SELECT position, aggregation_formula(count,'personnel_name') FROM agg_result ORDER BY position;

DROP TABLE agg_result;
