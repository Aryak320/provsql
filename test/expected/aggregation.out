\set ECHO none
remove_provenance

(1 row)
position|count|formula
Analyst|2 (*)|δ(Dave ⊕ Susan)
Director|1 (*)|δ(John)
Double agent|1 (*)|δ(Magdalen)
Field agent|1 (*)|δ(Ellen)
HR|1 (*)|δ(Nancy)
Janitor|1 (*)|δ(Paul)
(6 rows)
position|aggregation_formula
Analyst|sum{ (1 * Susan) , (1 * Dave) }
Director|sum{ (1 * John) }
Double agent|sum{ (1 * Magdalen) }
Field agent|sum{ (1 * Ellen) }
HR|sum{ (1 * Nancy) }
Janitor|sum{ (1 * Paul) }
(6 rows)
remove_provenance

(1 row)
position|aggregation_formula
Director|sum{ (1 * John) }
Double agent|sum{ (1 * Magdalen) }
Field agent|sum{ (1 * Ellen) }
HR|sum{ (1 * Nancy) }
Janitor|sum{ (1 * Paul) }
(5 rows)
remove_provenance

(1 row)
position|aggregation_formula
Director|sum{ (1 * John) }
Double agent|sum{ (1 * Magdalen) }
Field agent|sum{ (1 * Ellen) }
HR|sum{ (1 * Nancy) }
Janitor|sum{ (1 * Paul) }
(5 rows)
